homeassistant/util/unit_conversion.py

```diff
@@ -0,0 +1,141 @@
+"""Typing Helpers for Home Assistant."""
+from __future__ import annotations
+
+from numbers import Number
+from typing import overload
+
+from homeassistant.const import (
+    ENERGY_KILO_WATT_HOUR,
+    ENERGY_MEGA_WATT_HOUR,
+    ENERGY_WATT_HOUR,
+    POWER_KILO_WATT,
+    POWER_WATT,
+    PRESSURE_PA,
+    TEMP_CELSIUS,
+    UNIT_NOT_RECOGNIZED_TEMPLATE,
+    VOLUME_CUBIC_METERS,
+)
+
+from . import (
+    pressure as pressure_util,
+    temperature as temperature_util,
+    volume as volume_util,
+)
+
+
+class BaseUnitConverter:
+    """Define the format of a conversion utility."""
+
+    DEVICE_CLASS: str
+    NORMALIZED_UNIT: str
+    UNIT_CONVERSION: dict[str, float]
+    VALID_UNITS: tuple[str, ...]
+
+    @overload
+    @classmethod
+    def convert(
+        cls, value: float, from_unit: str, to_unit: str, bypass_checks: bool = False
+    ) -> float:
+        pass
+
+    @overload
+    @classmethod
+    def convert(
+        cls, value: None, from_unit: str, to_unit: str, bypass_checks: bool = False
+    ) -> None:
+        pass
```

Comment:
Instead of overloads, you can use a TypeVar with constraints for `value` and the return type.
```py
_ValueT = TypeVar("_ValueT", float, None)
```

Reply:
Good idea! done.

Reply:
Note: the previous implementation only supported `float` so I have reverted and kept for a follow-up PR.